add channels last for MaxPool2d

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51185 add channels last support for ConvTranspose2d
* #50573 add channels last support for PixelShuffle and PixelUnshuffle
* #50247 add channels last support for ChannelShuffle
* #49984 add channel last support for MaxUnpool2d
* #49821 add channels last for GroupNorm
* #49582 add channels last support for thnn_conv2d (non-dilated)
* #48920 add channels last for AdapativeMaxPool2d
* #48919 optimize channels last for BatchNorm2d on CPU
* #48918 add channels last support for AvgPool2d on CPU
* **#48917 add channels last for MaxPool2d**

max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

mingfeima committed:
add channels last for MaxPool2d

max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

[ghstack-poisoned]

mingfeima commented:
use this one to replace #42719


This patch adds channels last memory format support for `nn.MaxPool2d` on CPU, CL path is manually vectorized with `vec256` on dimension of C.

Performance result on  CPU Intel(R) Xeon(R) Platinum 8260 CPU @ 2.40GHz, 2*20 cores, comparing CL v.s. contiguous:
* single core inference (BS=1): **9.7x** speedup
* single socket inference (BS=1): **3.3x** speedup
* single socket inference (BS=128): **2.6x** speedup

input size is picked up from RN50. Use this [script](https://github.com/mingfeima/op_bench-py/blob/master/max_pool2d.py) to reproduce, `./run.sh max_pool2d.py`.

```bash
### (before)
### using OMP_NUM_THREADS=1
MaxPool2d(contiguous): input_size [1, 64, 112, 112] time: 5.317 ms

### using OMP_NUM_THREADS=20
MaxPool2d(contiguous): input_size [1, 64, 112, 112] time: 6.443 ms
MaxPool2d(contiguous): input_size [128, 64, 112, 112] time: 45.178 ms

### (after)
### using OMP_NUM_THREADS=1
MaxPool2d(contiguous):   input_size [1, 64, 112, 112] time: 5.304 ms
MaxPool2d(channels_last): input_size [1, 64, 112, 112] time: 0.544 ms

### using OMP_NUM_THREADS=20

MaxPool2d(contiguous):   input_size [1, 64, 112, 112] time: 0.336 ms
MaxPool2d(channels_last): input_size [1,   64, 112, 112] time: 0.102 ms
MaxPool2d(contiguous): input_size [128,   64, 112, 112] time: 42.141 ms
MaxPool2d(channels_last): input_size   [128, 64, 112, 112] time: 15.897 ms


```

VitalyFedyunin commented:
Fails internally on Android VR build:

```
Summary: 
[removed]/caffe2/aten/src/ATen/native/cpu/MaxPoolKernel.cpp:172:57: error: no member named 'isnan' in 'at::vec256::(anonymous namespace)::Vec256<float>'
            Vec mask = (val_vec > maxval_vec) | val_vec.isnan();
                                                ~~~~~~~ ^
stderr: [removed]/caffe2/aten/src/ATen/native/cpu/MaxPoolKernel.cpp:172:57: error: no member named 'isnan' in 'at::vec256::(anonymous namespace)::Vec256<float>'
            Vec mask = (val_vec > maxval_vec) | val_vec.isnan();
                                                ~~~~~~~ ^
 ** Summary of failures encountered during the build **
Rule [removed]
 FAILED because Command failed with exit code 1.
```

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

mingfeima commented:
> Fails internally on Android VR build:
> 
> ```
> Summary: 
> [removed]/caffe2/aten/src/ATen/native/cpu/MaxPoolKernel.cpp:172:57: error: no member named 'isnan' in 'at::vec256::(anonymous namespace)::Vec256<float>'
>             Vec mask = (val_vec > maxval_vec) | val_vec.isnan();
>                                                 ~~~~~~~ ^
> stderr: [removed]/caffe2/aten/src/ATen/native/cpu/MaxPoolKernel.cpp:172:57: error: no member named 'isnan' in 'at::vec256::(anonymous namespace)::Vec256<float>'
>             Vec mask = (val_vec > maxval_vec) | val_vec.isnan();
>                                                 ~~~~~~~ ^
>  ** Summary of failures encountered during the build **
> Rule [removed]
>  FAILED because Command failed with exit code 1.
> ```

Is there any guide or instructions that how i can build PyTorch on Android? @VitalyFedyunin

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

VitalyFedyunin commented on aten/src/ATen/cpu/vec256/vec256_base.h:
```diff
@@ -201,6 +201,17 @@ struct Vec256 {
     }
     return mask;
   }
+  Vec256<T> isnan() const {
```
Please add test here https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/test/vec256_test_all_types.cpp

mingfeima replied:
@VitalyFedyunin  test case added in `vec256_test_all_types.cpp`, please review!

VitalyFedyunin reviewed (commented):
Hello! I finished with OneDDN PRs and moving to this stack. Meanwhile can you please add vec256 test for new function.

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

VitalyFedyunin commented:
Code looks good, but as I can see you have dropped single frame dim()==3 optimization, can you please include rationale and benchmarks prooving new code is more effective (for both contiguous and channels last kernels)

mingfeima commented:
I will refresh this stack with Xeon Broadwell CPUs (AVX2) numbers, probably going to take a few days. Since we had removed all AVX2 machines from our internal cluster, need to get them back.

mingfeima commented:
#### update 1 - performance improvement on #51393
Run on Xeon Cascade CPU (AVX512), single socket, 20 cores.

```bash
(before):
maxpool NCHW 6.228899020403421 GB/s 2.8030045591815393 GFLOP/s
maxpool NHWC 4.517494221162063 GB/s 2.0328723995229283 GFLOP/s

(after):
maxpool NCHW 11.933086066468682 GB/s 5.369888729910908 GFLOP/s
maxpool NHWC 82.1337642586323 GB/s 36.960193916384526 GFLOP/s

```

@jamesr66a

mingfeima commented:
#### update2 - benchmark 3D input (tensor format CHW)

This patch removes dim() == 3 path, reason as follows:

The original parallelization pattern is like:
 - 3D input (CHW): parallel on dimension C;
 - 4D input (NCHW): nested omp loop: outer loop parallel on dimension N and inner loop still parallel on dimension C. Nested omp loops are somehow troublesome so PyTorch has a constraint that inner omp loop would go sequential. So actually 4D input will only parallel on outer dimension of N.
 - 4D input (NHWC): original code will treat NHWC as non-contiguous NCHW and call `.contiguous()` first. So NHWC is even slower than NCHW.
 
The new code will do dimension collapse on all dimensions, CHW or NCHW:
 - 3D input (CHW): parallel on all 3 dims of CHW;
 - 4D input (NCHW): parallel on all 4 dims of NCHW;
 - 4D input (NHWC): parallel on NHW and do vectorization on C.
 
For single core run, there would be no difference.
For multi core run, usually we should try to achieve more parallelism as we can. If we choose to pick to parallel on a single dimension and the size of that dimension is not big enough, then hardware resources can not be fully utilized.
#51393 is a good example, the original code will choose to parallel on N which is only 10 and thus only 10 cores will be used.

The following table updates performance compare of CHW and NCHW (N=1) to evaluate the influence of removing dim() = 3 path:
```bash
(before)
### using OMP_NUM_THREADS=1
MaxPool2d(contiguous): input_size [64, 112, 112] time: 5.260 ms
MaxPool2d(contiguous): input_size [1, 64, 112, 112] time: 5.245 ms
MaxPool2d(channels_last): input_size [1, 64, 112, 112] time: 7.790 ms

### using OMP_NUM_THREADS=20
MaxPool2d(contiguous): input_size [64, 112, 112] time: 0.415 ms
MaxPool2d(contiguous): input_size [1, 64, 112, 112] time: 6.422 ms
MaxPool2d(channels_last): input_size [1, 64, 112, 112] time: 6.599 ms

(after)
### using OMP_NUM_THREADS=1
MaxPool2d(contiguous): input_size [64, 112, 112] time: 5.227 ms
MaxPool2d(contiguous): input_size [1, 64, 112, 112] time: 5.231 ms
MaxPool2d(channels_last): input_size [1, 64, 112, 112] time: 0.531 ms

### using OMP_NUM_THREADS=20
MaxPool2d(contiguous): input_size [64, 112, 112] time: 0.343 ms
MaxPool2d(contiguous): input_size [1, 64, 112, 112] time: 0.344 ms
MaxPool2d(channels_last): input_size [1, 64, 112, 112] time: 0.048 ms
```

we can see that
 - for single core run, the new code has same perf as original code.
 - for multi core run, the new code is better (0.343 ms v.s. 0.415 ms). Reason is that the original code will parallel on dim C (size of 64), and on a 20-core CPU, it will has a chunk_size of 4 (see: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelOpenMP.h#L43) and use only 16 cores. The new code will parallel on CHW (64\*112\*112), so it is pretty sure that all 20 cores will be used.
 
if we are comparing a CHW tensor of [3, 224, 224] on multi-core runs, the improvements would be even bigger.

mingfeima commented:
#### update3 - AVX2 CPU benchmark result:
* **avx2 CPU**: Xeon(R) CPU E5-2680 v3 @ 2.50GHz, dual sockets, 12 cores per socket. Codename **Haswell**, [spec](https://ark.intel.com/content/www/us/en/ark/products/81908/intel-xeon-processor-e5-2680-v3-30m-cache-2-50-ghz.html)

```bash
(before)
### using OMP_NUM_THREADS=1
MaxPool2d(contiguous): input_size [64, 112, 112] time: 7.213 ms
MaxPool2d(contiguous): input_size [1, 64, 112, 112] time: 7.137 ms
MaxPool2d(channels_last): input_size [1, 64, 112, 112] time: 8.980 ms

### using OMP_NUM_THREADS=12
MaxPool2d(contiguous): input_size [64, 112, 112] time: 0.779 ms
MaxPool2d(contiguous): input_size [1, 64, 112, 112] time: 7.349 ms
MaxPool2d(channels_last): input_size [1, 64, 112, 112] time: 7.662 ms
MaxPool2d(contiguous): input_size [128, 64, 112, 112] time: 101.917 ms
MaxPool2d(channels_last): input_size [128, 64, 112, 112] time: 141.047 ms


(after)
### using OMP_NUM_THREADS=1
MaxPool2d(contiguous): input_size [64, 112, 112] time: 7.253 ms
MaxPool2d(contiguous): input_size [1, 64, 112, 112] time: 7.187 ms
MaxPool2d(channels_last): input_size [1, 64, 112, 112] time: 0.868 ms

### using OMP_NUM_THREADS=12
MaxPool2d(contiguous): input_size [64, 112, 112] time: 0.787 ms
MaxPool2d(contiguous): input_size [1, 64, 112, 112] time: 0.771 ms
MaxPool2d(channels_last): input_size [1, 64, 112, 112] time: 0.239 ms
MaxPool2d(contiguous): input_size [128, 64, 112, 112] time: 102.118 ms
MaxPool2d(channels_last): input_size [128, 64, 112, 112] time: 27.572 ms
```

We can see that the optimization is equally effective on AVX2 CPU.

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

VitalyFedyunin reviewed (approved):
Thanks for the follow-up benchmarking and explanation

VitalyFedyunin commented:
Getting bunch of internal build errors:
```
Summary: 
ld.lld: error: undefined symbol: at::native::DispatchStub<void (*)(at::Tensor&, at::Tensor&, at::Tensor const&, int, int, int, int, int, int, int, int), at::native::max_pool2d_kernel>::DEFAULT
ld.lld: error: undefined symbol: at::native::DispatchStub<void (*)(at::Tensor&, at::Tensor const&, at::Tensor const&), at::native::max_pool2d_backward_kernel>::DEFAULT
ld.lld: error: undefined symbol: at::native::DispatchStub<void (*)(at::Tensor&, at::Tensor&, at::Tensor const&, int, int, int, int, int, int, int, int), at::native::max_pool2d_kernel>::DEFAULT
ld.lld: error: undefined symbol: at::native::DispatchStub<void (*)(at::Tensor&, at::Tensor const&, at::Tensor const&), at::native::max_pool2d_backward_kernel>::DEFAULT
```
trying to figure out what is going on now

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

jgong5 commented:
Hi @VitalyFedyunin : any luck troubleshooting the internal build failures?

VitalyFedyunin commented:
We made some adjustments to the build system, which should unlock landing of this stack. Could you please rebase so I can try tomorrow. Thanks.

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

mingfeima commented:
Hi @VitalyFedyunin, the stack has been rebased! Please review!
Sorry for the delay, got occupied by some other task.

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

VitalyFedyunin commented:
Sorry, merge conflicts with just landed https://github.com/pytorch/pytorch/pull/54898 , please rebase again

mingfeima committed:
Update on "add channels last for MaxPool2d"


max_pool2d channels last support forward path

max_pool2d channels last support backward path

vectorize channels last forward path

rename the header file

fix windows build

combine PoolingKernel.h into Pool.h

add data type check

loosen test_max_pool2d_nhwc to cover device CPU

Differential Revision: [D25399470](https://our.internmc.facebook.com/intern/diff/D25399470)

[ghstack-poisoned]

mingfeima commented:
@VitalyFedyunin updated!

agolynski commented:
Seems like this broke ios build MaxPoolKernel.cpp

VitalyFedyunin commented:
Broken build url https://app.circleci.com/pipelines/github/pytorch/pytorch/295279/workflows/5f586a11-d066-46a7-a085-9b72258ce6bc/jobs/12076475

mingfeima commented:
> Broken build url https://app.circleci.com/pipelines/github/pytorch/pytorch/295279/workflows/5f586a11-d066-46a7-a085-9b72258ce6bc/jobs/12076475

working on it.

mingfeima commented:
I can't reproduce the build failure on this patch on iOS.
I followed the instruction from https://pytorch.org/mobile/ios/#build-libtorch-for-arm64-devices
```bash
BUILD_PYTORCH_MOBILE=1 IOS_PLATFORM=SIMULATOR ./scripts/build_ios.sh
```

I built on both iOS and MacOS and both succeeded.
My local environment is
```
-- Toolchain using default iOS SDK: /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS14.4.sdk
-- The CXX compiler identification is AppleClang 12.0.0.12000032
-- The C compiler identification is AppleClang 12.0.0.12000032
```

About the failure from https://app.circleci.com/pipelines/github/pytorch/pytorch/295279/workflows/5f586a11-d066-46a7-a085-9b72258ce6bc/jobs/12076475

The thing that goes wrong is `MaxPoolKernel.cpp` has been compiled twice

1. the first time succeeded, `MaxPoolKernel.cpp` is a kernel file from ATen/native/cpu directory and it has been compiled successfully (with .DEFAULT) at
```
Apr 02 16:27:52 [ 84%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/MaxPoolKernel.cpp.DEFAULT.cpp.o
```

2. second time goes wrong, this should not exist
```
Apr 02 16:33:41 [ 93%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/MaxPoolKernel.cpp.o
```

I don't understand why both `MaxPoolKernel.cpp` and `MaxPoolKernel.cpp.DEFAULT.cpp` are compiled.

@VitalyFedyunin  About the failed CI, is there any special build recipe for it?

malfet commented on aten/src/ATen/native/Pool.h:
```diff
@@ -1,14 +1,20 @@
+    int kW, int kH, int dW, int dH, int padW, int padH, int dilationW, int dilationH);
+using max_pool2d_backward_fn = void(*)(Tensor& grad_input, const Tensor& grad_output, const Tensor& indices);
+
+DECLARE_DISPATCH(max_pool2d_fn, max_pool2d_kernel);
```
I think this PR misses a `DEFINE_DISPATCH` call in one of .cpp files

malfet commented on aten/src/ATen/cpu/vec256/vec256_float_neon.h:
```diff
@@ -282,6 +282,19 @@ template <> class Vec256<float> {
+    __at_align32__ float tmp[size()];
+    __at_align32__ float res[size()];
+    store(tmp);
+    for (int64_t i = 0; i < size(); i++) {
```
Why `int64_t`, when `decltype(size())` is int?

mingfeima replied:
int should be enough.

malfet commented on aten/src/ATen/native/cpu/MaxPoolKernel.cpp:
```diff
@@ -0,0 +1,361 @@
+    int64_t ow = 0;
+    data_index_init(begin, c, channels, oh, output_height, ow, output_width);
+
+    for (int64_t i = begin; i < end; i++) {
```
Can `c10::irange` be used here?

VitalyFedyunin commented:
@mingfeima lets try to figure out what is going on here. Could you please create this change as separate PR (do NOT abandon stack) and add `ci-all` tag to it.

Also please add `aten/src/ATen/native/cpu/MaxPoolKernel.cpp` into the `tools/build_variables.bzl`

malfet commented:
> I can't reproduce the build failure on this patch on iOS.
> I followed the instruction from https://pytorch.org/mobile/ios/#build-libtorch-for-arm64-devices
This only happens when Lite interpreter is build, i.e. when `BUILD_LITE_INTERPRETER=1`
And to fix the problem, add `MaxPoolKernel.cpp` to 
https://github.com/pytorch/pytorch/blob/bbdb37b93d82b17f2f4fabd032398a2b5461bfb8/tools/build_variables.bzl#L759

@cccclai Can you explain why this must be an explicit list instead of glob?

cccclai commented:
> ```
> Apr 02 16:33:41 [ 93%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/MaxPoolKernel.cpp.o
> ```

Yes the recipe is to build lite interpreter is here: https://pytorch.org/tutorials/prototype/lite_interpreter.html. To build for ios, the command is: 
```
BUILD_PYTORCH_MOBILE=1 IOS_PLATFORM=SIMULATOR BUILD_LITE_INTERPRETER=1 ./scripts/build_ios.sh
```

cccclai commented:
> > I can't reproduce the build failure on this patch on iOS.
> > I followed the instruction from https://pytorch.org/mobile/ios/#build-libtorch-for-arm64-devices
> 
> This only happens when Lite interpreter is build, i.e. when `BUILD_LITE_INTERPRETER=1`
> And to fix the problem, add `MaxPoolKernel.cpp` to
> https://github.com/pytorch/pytorch/blob/bbdb37b93d82b17f2f4fabd032398a2b5461bfb8/tools/build_variables.bzl#L759
> 
> @cccclai Can you please explain why this must be an explicit list instead of glob?

The main reason is to pick the files needed for pytorch edge, such that we can build a library with smaller size (lite interpreter). Some resources from aten are not needed when deploying for mobile.
